File: crates/quil-store/src/worker.rs
//! Worker registry persistence. Mirrors Go's
//! `node/store/worker.go::PebbleWorkerStore` so the operator's
//! `manually_managed` flag and the worker→filter binding survive
//! node restarts.
//!
//! Layout:
//!   * `[WORKER, WORKER_BY_CORE, core_id_be_u64]` → encoded
//!     `PersistedWorkerInfo`
//!   * `[WORKER, WORKER_BY_FILTER, filter…]` → 8-byte BE core_id
//!     (secondary index; lets Go-style lookups by filter resolve
//!     to the primary record)
//!
//! Encoding mirrors Go's `encodeWorkerInfo` field order so a
//! Rust-written record could in principle be read by Go and vice
//! versa, modulo the listen-multiaddr fields which we don't track
//! (they're empty strings in the encoding — same on-disk shape).

use std::sync::Arc;

use quil_types::error::{QuilError, Result};
use quil_types::store::{PersistedWorkerInfo, WorkerStore};

use crate::encoding::{WORKER, WORKER_BY_CORE, WORKER_BY_FILTER};

pub struct RocksWorkerStore {
    db: Arc<rocksdb::DB>,
}

impl RocksWorkerStore {
    pub fn new(db: Arc<rocksdb::DB>) -> Self {
        Self { db }
    }
}

fn worker_key(core_id: u32) -> [u8; 10] {
    let mut k = [0u8; 10];
    k[0] = WORKER;
    k[1] = WORKER_BY_CORE;
    k[2..10].copy_from_slice(&(core_id as u64).to_be_bytes());
    k
}

fn worker_by_filter_key(filter: &[u8]) -> Vec<u8> {
    let mut k = Vec::with_capacity(2 + filter.len());
    k.push(WORKER);
    k.push(WORKER_BY_FILTER);
    k.extend_from_slice(filter);
    k
}

fn encode(info: &PersistedWorkerInfo) -> Vec<u8> {
    // Mirrors Go's `encodeWorkerInfo`:
    //   coreId(8) + totalStorage(8) + automatic(1) + allocated(1)
    //   + listenMultiaddrLen(2) + listenMultiaddr
    //   + streamListenMultiaddrLen(2) + streamListenMultiaddr
    //   + filterLen(2) + filter
    //   + pendingFilterFrame(8)
    //   + manuallyManaged(1)
    let filter_len = info.filter.len() as u16;
    let total_len = 8 + 8 + 1 + 1 + 2 + 2 + 2 + filter_len as usize + 8 + 1;
    let mut out = Vec::with_capacity(total_len);
    out.extend_from_slice(&(info.core_id as u64).to_be_bytes());
    // total_storage — not tracked locally; emit 0 for forward compat.
    out.extend_from_slice(&0u64.to_be_bytes());
    // automatic = !manually_managed (Go's terminology)
    out.push(if info.manually_managed { 0 } else { 1 });
    out.push(if info.allocated { 1 } else { 0 });
    // listen / stream multiaddrs — empty in this codepath.
    out.extend_from_slice(&0u16.to_be_bytes());
    out.extend_from_slice(&0u16.to_be_bytes());
    out.extend_from_slice(&filter_len.to_be_bytes());
    out.extend_from_slice(&info.filter);
    out.extend_from_slice(&info.pending_filter_frame.to_be_bytes());
    out.push(if info.manually_managed { 1 } else { 0 });
    out
}

fn decode(data: &[u8]) -> Result<PersistedWorkerInfo> {
    if data.len() < 24 {
        return Err(QuilError::Store("worker info: too short".into()));
    }
    let mut o = 0;
    let core_id_u64 = u64::from_be_bytes(data[o..o + 8].try_into().unwrap());
    o += 8;
    let _total_storage = u64::from_be_bytes(data[o..o + 8].try_into().unwrap());
    o += 8;
    // automatic byte (consumed via the trailing manually_managed
    // byte; Go writes both for compat).
    let _automatic = data[o];
    o += 1;
    let allocated = data[o] == 1;
    o += 1;
    if o + 2 > data.len() {
        return Err(QuilError::Store("worker info: truncated listen len".into()));
    }
    let listen_len = u16::from_be_bytes(data[o..o + 2].try_into().unwrap()) as usize;
    o += 2;
    if o + listen_len > data.len() {
        return Err(QuilError::Store("worker info: truncated listen".into()));
    }
    o += listen_len;
    if o + 2 > data.len() {
        return Err(QuilError::Store("worker info: truncated stream len".into()));
    }
    let stream_len = u16::from_be_bytes(data[o..o + 2].try_into().unwrap()) as usize;
    o += 2;
    if o + stream_len > data.len() {
        return Err(QuilError::Store("worker info: truncated stream".into()));
    }
    o += stream_len;
    if o + 2 > data.len() {
        return Err(QuilError::Store("worker info: truncated filter len".into()));
    }
    let filter_len = u16::from_be_bytes(data[o..o + 2].try_into().unwrap()) as usize;
    o += 2;
    if o + filter_len > data.len() {
        return Err(QuilError::Store("worker info: truncated filter".into()));
    }
    let filter = data[o..o + filter_len].to_vec();
    o += filter_len;
    let pending_filter_frame = if o + 8 <= data.len() {
        let v = u64::from_be_bytes(data[o..o + 8].try_into().unwrap());
        o += 8;
        v
    } else {
        0
    };
    let manually_managed = data.get(o).copied().unwrap_or(0) == 1;

    Ok(PersistedWorkerInfo {
        core_id: core_id_u64 as u32,
        filter,
        manually_managed,
        allocated,
        pending_filter_frame,
    })
}

impl WorkerStore for RocksWorkerStore {
    fn get_worker(&self, core_id: u32) -> Result<Option<PersistedWorkerInfo>> {
        let key = worker_key(core_id);
        let raw = self
            .db
            .get(&key)
            .map_err(|e| QuilError::Store(format!("get_worker: {}", e)))?;
        match raw {
            Some(bytes) => Ok(Some(decode(&bytes)?)),
            None => Ok(None),
        }
    }

    fn put_worker(&self, worker: &PersistedWorkerInfo) -> Result<()> {
        // Clear any stale by-filter index entry first.
        if let Some(existing) = self.get_worker(worker.core_id)? {
            if !existing.filter.is_empty() && existing.filter != worker.filter {
                let _ = self.db.delete(worker_by_filter_key(&existing.filter));
            }
        }
        let pri_key = worker_key(worker.core_id);
        let payload = encode(worker);
        self.db
            .put(&pri_key, &payload)
            .map_err(|e| QuilError::Store(format!("put_worker: {}", e)))?;
        if !worker.filter.is_empty() {
            let idx_key = worker_by_filter_key(&worker.filter);
            self.db
                .put(&idx_key, &(worker.core_id as u64).to_be_bytes())
                .map_err(|e| QuilError::Store(format!("put_worker: {}", e)))?;
        }
        Ok(())
    }

    fn delete_worker(&self, core_id: u32) -> Result<()> {
        if let Some(existing) = self.get_worker(core_id)? {
            if !existing.filter.is_empty() {
                let _ = self.db.delete(worker_by_filter_key(&existing.filter));
            }
        }
        self.db
            .delete(&worker_key(core_id))
            .map_err(|e| QuilError::Store(format!("delete_worker: {}", e)))
    }

    fn range_workers(&self) -> Result<Vec<PersistedWorkerInfo>> {
        let lower = [WORKER, WORKER_BY_CORE, 0x00, 0, 0, 0, 0, 0, 0, 0];
        let upper = [WORKER, WORKER_BY_CORE, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF];
        let mut read_opts = rocksdb::ReadOptions::default();
        read_opts.set_iterate_lower_bound(lower.to_vec());
        read_opts.set_iterate_upper_bound(upper.to_vec());
        let iter = self.db.iterator_opt(rocksdb::IteratorMode::Start, read_opts);
        let mut out = Vec::new();
        for item in iter {
            let (_k, v) = item
                .map_err(|e| QuilError::Store(format!("range_workers: {}", e)))?;
            out.push(decode(&v)?);
        }
        out.sort_by_key(|w| w.core_id);
        Ok(out)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::RocksDb;
    use tempfile::TempDir;

    fn fresh_store() -> (TempDir, RocksWorkerStore) {
        let dir = TempDir::new().unwrap();
        let db = RocksDb::open(dir.path()).unwrap();
        (dir, RocksWorkerStore::new(db.inner()))
    }

    #[test]
    fn put_get_round_trips() {
        let (_dir, s) = fresh_store();
        let w = PersistedWorkerInfo {
            core_id: 3,
            filter: vec![0xA1; 33],
            manually_managed: true,
            allocated: true,
            pending_filter_frame: 12345,
        };
        s.put_worker(&w).unwrap();
        let back = s.get_worker(3).unwrap().unwrap();
        assert_eq!(back, w);
    }

    #[test]
    fn range_returns_sorted() {
        let (_dir, s) = fresh_store();
        for id in [4u32, 1, 7] {
            s.put_worker(&PersistedWorkerInfo {
                core_id: id,
                filter: vec![id as u8; 8],
                manually_managed: id == 1,
                allocated: true,
                pending_filter_frame: 0,
            })
            .unwrap();
        }
        let all = s.range_workers().unwrap();
        let ids: Vec<u32> = all.iter().map(|w| w.core_id).collect();
        assert_eq!(ids, vec![1, 4, 7]);
        assert!(all[0].manually_managed);
        assert!(!all[1].manually_managed);
    }

    #[test]
    fn delete_removes_primary_and_filter_index() {
        let (_dir, s) = fresh_store();
        let w = PersistedWorkerInfo {
            core_id: 5,
            filter: vec![9, 9, 9],
            manually_managed: false,
            allocated: true,
            pending_filter_frame: 0,
        };
        s.put_worker(&w).unwrap();
        s.delete_worker(5).unwrap();
        assert!(s.get_worker(5).unwrap().is_none());
    }
}

File: node/store/worker.go
package store

import (
	"encoding/binary"
	"fmt"
	"slices"

	"github.com/cockroachdb/pebble/v2"
	"github.com/pkg/errors"
	"go.uber.org/zap"
	"source.quilibrium.com/quilibrium/monorepo/types/store"
)

var _ store.WorkerStore = (*PebbleWorkerStore)(nil)

type PebbleWorkerStore struct {
	db     store.KVDB
	logger *zap.Logger
}

func NewPebbleWorkerStore(
	db store.KVDB,
	logger *zap.Logger,
) *PebbleWorkerStore {
	return &PebbleWorkerStore{
		db,
		logger,
	}
}

func workerKey(coreId uint) []byte {
	key := []byte{WORKER, WORKER_BY_CORE}
	coreIdBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(coreIdBytes, uint64(coreId))
	key = append(key, coreIdBytes...)
	return key
}

func workerByFilterKey(filter []byte) []byte {
	key := []byte{WORKER, WORKER_BY_FILTER}
	key = append(key, filter...)
	return key
}

func (p *PebbleWorkerStore) NewTransaction(indexed bool) (
	store.Transaction,
	error,
) {
	return p.db.NewBatch(indexed), nil
}

func (p *PebbleWorkerStore) GetWorker(coreId uint) (*store.WorkerInfo, error) {
	data, closer, err := p.db.Get(workerKey(coreId))
	if err != nil {
		if errors.Is(err, pebble.ErrNotFound) {
			return nil, store.ErrNotFound
		}
		return nil, errors.Wrap(err, "get worker")
	}

	copied := slices.Clone(data)
	closer.Close()

	return decodeWorkerInfo(copied)
}

func (p *PebbleWorkerStore) GetWorkerByFilter(filter []byte) (
	*store.WorkerInfo,
	error,
) {
	if len(filter) == 0 {
		return nil, errors.Wrap(
			errors.New("filter cannot be empty"),
			"get worker by filter",
		)
	}

	data, closer, err := p.db.Get(workerByFilterKey(filter))
	if err != nil {
		if errors.Is(err, pebble.ErrNotFound) {
			return nil, store.ErrNotFound
		}
		return nil, errors.Wrap(err, "get worker by filter")
	}
	copied := slices.Clone(data)
	closer.Close()

	if len(copied) < 8 {
		return nil, errors.Wrap(
			fmt.Errorf("index payload too short: %d", len(copied)),
			"get worker by filter",
		)
	}
	coreId := binary.BigEndian.Uint64(copied[:8])
	return p.GetWorker(uint(coreId))
}

func (p *PebbleWorkerStore) PutWorker(
	txn store.Transaction,
	worker *store.WorkerInfo,
) error {
	// Check if worker already exists to clean up old filter index if needed
	existingWorker, err := p.GetWorker(worker.CoreId)
	if err == nil && existingWorker != nil {
		// Delete old filter index if it exists and is different
		if len(existingWorker.Filter) > 0 &&
			string(existingWorker.Filter) != string(worker.Filter) {
			if err := txn.Delete(
				workerByFilterKey(existingWorker.Filter),
			); err != nil {
				return errors.Wrap(err, "put worker")
			}
		}
	}

	data, err := encodeWorkerInfo(worker)
	if err != nil {
		return errors.Wrap(err, "put worker")
	}

	if err := txn.Set(workerKey(worker.CoreId), data); err != nil {
		return errors.Wrap(err, "put worker")
	}

	// Only set filter index if filter is not empty
	if len(worker.Filter) > 0 {
		coreIdBytes := make([]byte, 8)
		binary.BigEndian.PutUint64(coreIdBytes, uint64(worker.CoreId))
		if err := txn.Set(
			workerByFilterKey(worker.Filter),
			coreIdBytes,
		); err != nil {
			return errors.Wrap(err, "put worker")
		}
	}

	return nil
}

func (p *PebbleWorkerStore) DeleteWorker(
	txn store.Transaction,
	coreId uint,
) error {
	worker, err := p.GetWorker(coreId)
	if err != nil {
		return errors.Wrap(err, "delete worker")
	}

	if err := txn.Delete(workerKey(coreId)); err != nil {
		return errors.Wrap(err, "delete worker")
	}

	// Only delete filter index if filter is not empty
	if len(worker.Filter) > 0 {
		if err := txn.Delete(workerByFilterKey(worker.Filter)); err != nil {
			return errors.Wrap(err, "delete worker")
		}
	}

	return nil
}

func (p *PebbleWorkerStore) RangeWorkers() ([]*store.WorkerInfo, error) {
	iter, err := p.db.NewIter(
		[]byte{WORKER, WORKER_BY_CORE, 0x00},
		[]byte{WORKER, WORKER_BY_CORE, 0xFF},
	)
	if err != nil {
		return nil, errors.Wrap(err, "range workers")
	}
	defer iter.Close()

	var workers []*store.WorkerInfo
	for iter.First(); iter.Valid(); iter.Next() {
		val := slices.Clone(iter.Value())
		worker, err := decodeWorkerInfo(val)
		if err != nil {
			return nil, errors.Wrap(err, "range workers")
		}
		workers = append(workers, worker)
	}

	return workers, nil
}

func encodeWorkerInfo(worker *store.WorkerInfo) ([]byte, error) {
	listenMultiaddrLen := uint16(len(worker.ListenMultiaddr))
	streamListenMultiaddrLen := uint16(len(worker.StreamListenMultiaddr))
	filterLen := uint16(len(worker.Filter))

	// totalLen = coreId(8) + totalStorage(8) + automatic(1) + allocated(1)
	//   + 2 + listen + 2 + stream + 2 + filter + pendingFilterFrame(8)
	//   + manuallyManaged(1)
	totalLen := 8 + 8 + 1 + 1 + 2 + int(listenMultiaddrLen) + 2 +
		int(streamListenMultiaddrLen) + 2 + int(filterLen) + 8 + 1
	data := make([]byte, totalLen)

	offset := 0
	binary.BigEndian.PutUint64(data[offset:], uint64(worker.CoreId))
	offset += 8

	binary.BigEndian.PutUint64(data[offset:], uint64(worker.TotalStorage))
	offset += 8

	if worker.Automatic {
		data[offset] = 1
	} else {
		data[offset] = 0
	}
	offset += 1

	if worker.Allocated {
		data[offset] = 1
	} else {
		data[offset] = 0
	}
	offset += 1

	binary.BigEndian.PutUint16(data[offset:], listenMultiaddrLen)
	offset += 2
	copy(data[offset:], worker.ListenMultiaddr)
	offset += int(listenMultiaddrLen)

	binary.BigEndian.PutUint16(data[offset:], streamListenMultiaddrLen)
	offset += 2
	copy(data[offset:], worker.StreamListenMultiaddr)
	offset += int(streamListenMultiaddrLen)

	binary.BigEndian.PutUint16(data[offset:], filterLen)
	offset += 2
	copy(data[offset:], worker.Filter)
	offset += int(filterLen)

	binary.BigEndian.PutUint64(data[offset:], worker.PendingFilterFrame)
	offset += 8

	if worker.ManuallyManaged {
		data[offset] = 1
	}

	return data, nil
}

func decodeWorkerInfo(data []byte) (*store.WorkerInfo, error) {
	if len(data) < 24 {
		return nil, errors.New("invalid worker info data: too short")
	}

	offset := 0
	if offset+8 > len(data) {
		return nil, errors.New("truncated coreId")
	}
	coreId := binary.BigEndian.Uint64(data[offset:])
	offset += 8

	if offset+8 > len(data) {
		return nil, errors.New("truncated totalStorage")
	}
	totalStorage := binary.BigEndian.Uint64(data[offset:])
	offset += 8

	if offset+1 > len(data) {
		return nil, errors.New("truncated automatic flag")
	}
	automatic := data[offset] == 1
	offset += 1

	if offset+1 > len(data) {
		return nil, errors.New("truncated allocated flag")
	}
	allocated := data[offset] == 1
	offset += 1

	if offset+2 > len(data) {
		return nil, errors.New("truncated listenMultiaddr length")
	}

	listenMultiaddrLen := int(binary.BigEndian.Uint16(data[offset:]))
	offset += 2
	if listenMultiaddrLen < 0 || offset+listenMultiaddrLen > len(data) {
		return nil, errors.New("invalid listen multiaddr length")
	}

	listenMultiaddr := string(data[offset : offset+listenMultiaddrLen])
	offset += listenMultiaddrLen

	if offset+2 > len(data) {
		return nil, errors.New("truncated streamListenMultiaddr length")
	}

	streamListenMultiaddrLen := int(binary.BigEndian.Uint16(data[offset:]))
	offset += 2
	if streamListenMultiaddrLen < 0 ||
		offset+streamListenMultiaddrLen > len(data) {
		return nil, errors.New("invalid stream listen multiaddr length")
	}

	streamListenMultiaddr := string(
		data[offset : offset+streamListenMultiaddrLen],
	)
	offset += streamListenMultiaddrLen

	if offset+2 > len(data) {
		return nil, errors.New("truncated filter length")
	}

	filterLen := int(binary.BigEndian.Uint16(data[offset:]))
	offset += 2
	if filterLen < 0 || offset+filterLen > len(data) {
		return nil, errors.New("invalid filter length")
	}

	filter := make([]byte, filterLen)
	copy(filter, data[offset:offset+filterLen])
	offset += filterLen

	var pendingFrame uint64
	if offset+8 <= len(data) {
		pendingFrame = binary.BigEndian.Uint64(data[offset:])
		offset += 8
	}

	var manuallyManaged bool
	if offset+1 <= len(data) {
		manuallyManaged = data[offset] == 1
	}

	return &store.WorkerInfo{
		CoreId:                uint(coreId),
		ListenMultiaddr:       listenMultiaddr,
		StreamListenMultiaddr: streamListenMultiaddr,
		Filter:                filter,
		TotalStorage:          uint(totalStorage),
		Automatic:             automatic,
		Allocated:             allocated,
		PendingFilterFrame:    pendingFrame,
		ManuallyManaged:       manuallyManaged,
	}, nil
}
